panic: Unexpected: got 3 chunks from chunk.Add

```
panic: Unexpected: got 3 chunks from chunk.Add

goroutine 1087224 [running]:
panic(0xab7ac0, 0xc44a872220)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/weaveworks/cortex/ingester.(*memorySeries).add(0xc4298e2f00, 0x15993e5cfd1, 0x40936e99cff75fe2, 0x2ca8153190dfee9c, 0x0)
	/go/src/github.com/weaveworks/cortex/ingester/series.go:93 +0x49f
github.com/weaveworks/cortex/ingester.(*Ingester).append(0xc420221440, 0x7fee113f0840, 0xc44881e7e0, 0xc4452ac5e0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/ingester/ingester.go:236 +0x2b3
github.com/weaveworks/cortex/ingester.(*Ingester).Push(0xc420221440, 0x7fee113f0840, 0xc44881e7e0, 0xc44e01ea60, 0xc4452aa1c0, 0xc43ff97160, 0x1)
	/go/src/github.com/weaveworks/cortex/ingester/ingester.go:199 +0x79
github.com/weaveworks/cortex._Ingester_Push_Handler.func1(0x7fee113f0840, 0xc44881e7e0, 0xb34f60, 0xc44e01ea60, 0x7fee113f0840, 0xc44881e7e0, 0x115cd80, 0xc44881e7b0)
	/go/src/github.com/weaveworks/cortex/cortex.pb.go:421 +0xa2
github.com/weaveworks/cortex/util/middleware.ServerUserHeaderInterceptor(0x7fee113f0840, 0xc44881e7b0, 0xb34f60, 0xc44e01ea60, 0xc44cbc3fc0, 0xc44cbc3fe0, 0x7fee113f0840, 0xc44881e7b0, 0xaf8f80, 0x11dd748)
	/go/src/github.com/weaveworks/cortex/util/middleware/auth.go:44 +0x15a
github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x7fee113f0840, 0xc44881e7b0, 0xb34f60, 0xc44e01ea60, 0x7fee113f0840, 0xc44881e7b0, 0x1160760, 0x11dd748)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware/chain.go:21 +0x66
github.com/weaveworks/cortex/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x7fee113f0840, 0xc44881e7b0, 0xb34f60, 0xc44e01ea60, 0xc44cbc3fc0, 0xc4452ac000, 0x0, 0x0, 0x1153780, 0xc420128090)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:56 +0x39d
github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xffffffff, 0x0, 0xb, 0xc44d1877e0)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware/chain.go:21 +0x66
github.com/weaveworks/cortex/util/middleware.ServerInstrumentInterceptor.func1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xc44cbc3fc0, 0xc4452ac020, 0xc44d187878, 0x54b862, 0x5877c825, 0x26251df)
	/go/src/github.com/weaveworks/cortex/util/middleware/instrumentation.go:15 +0x9e
github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xc44d187900, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware/chain.go:21 +0x66
github.com/weaveworks/cortex/util/middleware.ServerLoggingInterceptor.func1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xc44cbc3fc0, 0xc4452ac040, 0xb55400, 0xc4452ac060, 0xc44cbc3fc0, 0xc4452ac060)
	/go/src/github.com/weaveworks/cortex/util/middleware/logging.go:17 +0x9d
github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware.ChainUnaryServer.func1.1.1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xc42000c340, 0xc44e01c990, 0xc44d187a70, 0x4159e8)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware/chain.go:21 +0x66
github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware.ChainUnaryServer.func1(0x7fee113f0840, 0xc43f88e300, 0xb34f60, 0xc44e01ea60, 0xc44cbc3fc0, 0xc44cbc3fe0, 0x52bb, 0xc43f64ab18, 0x0, 0xc43f64ab08)
	/go/src/github.com/weaveworks/cortex/vendor/github.com/mwitkow/go-grpc-middleware/chain.go:28 +0x118
github.com/weaveworks/cortex._Ingester_Push_Handler(0xbab640, 0xc420221440, 0x7fee113f0840, 0xc43f88e300, 0xc449ce3560, 0xc4202baf00, 0x0, 0x0, 0x50, 0xc41ef09ed4)
	/go/src/github.com/weaveworks/cortex/cortex.pb.go:423 +0x166
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc42000c300, 0x115f9a0, 0xc4201d00c0, 0xc44b922f00, 0xc4202dacf0, 0x1149e40, 0xc43f88e420, 0x0, 0x0)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/server.go:708 +0xbd8
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*Server).handleStream(0xc42000c300, 0x115f9a0, 0xc4201d00c0, 0xc44b922f00, 0xc43f88e420)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/server.go:892 +0x6a4
github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4203f44f0, 0xc42000c300, 0x115f9a0, 0xc4201d00c0, 0xc44b922f00)
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/server.go:468 +0xab
created by github.com/weaveworks/cortex/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/weaveworks/cortex/vendor/google.golang.org/grpc/server.go:469 +0xa3
```

--- a/ingester/series.go
+++ b/ingester/series.go
@@ -1,7 +1,6 @@
 package ingester
 
 import (
-	"fmt"
 	"sort"
 
 	"github.com/prometheus/client_golang/prometheus"
@@ -72,7 +71,7 @@ func (s *memorySeries) add(v model.SamplePair) error {
 	}
 
 	if len(s.chunkDescs) == 0 || s.headChunkClosed {
-		newHead := newDesc(chunk.New(), v.Timestamp)
+		newHead := newDesc(chunk.New(), v.Timestamp, v.Timestamp)
 		s.chunkDescs = append(s.chunkDescs, newHead)
 		s.headChunkClosed = false
 	}
@@ -81,16 +80,21 @@ func (s *memorySeries) add(v model.SamplePair) error {
 	if err != nil {
 		return err
 	}
-	s.head().C = chunks[0]
 
-	switch len(chunks) {
-	case 1: // noop, sample added to the chunk
-	case 2:
-		// new overflow chunk - in practice there is only ever 1 overflow chunk,
-		// thats how the chunk code is constructed.
-		s.chunkDescs = append(s.chunkDescs, newDesc(chunks[1], v.Timestamp))
-	default:
-		panic(fmt.Sprintf("Unexpected: got %d chunks from chunk.Add", len(chunks)))
+	// If we get a single chunk result, then just replace the head chunk with it
+	// (no need to update first/last time).  Otherwise, we'll need to update first
+	// and last time.
+	if len(chunks) == 1 {
+		s.head().C = chunks[0]
+	} else {
+		s.chunkDescs = s.chunkDescs[:len(s.chunkDescs)-1]
+		for _, c := range chunks {
+			lastTime, err := c.NewIterator().LastTimestamp()
+			if err != nil {
+				return err
+			}
+			s.chunkDescs = append(s.chunkDescs, newDesc(c, c.FirstTime(), lastTime))
+		}
 	}
 
 	return nil
@@ -159,11 +163,11 @@ type desc struct {
 	LastTime  model.Time  // Populated at creation & on append.
 }
 
-func newDesc(c chunk.Chunk, firstTime model.Time) *desc {
+func newDesc(c chunk.Chunk, firstTime model.Time, lastTime model.Time) *desc {
 	return &desc{
 		C:         c,
 		FirstTime: firstTime,
-		LastTime:  firstTime,
+		LastTime:  lastTime,
 	}
 }
 

